### PR TITLE
Detailed documentation of new API design and Data model

### PR DESCRIPTION
**Pull request by eliascodes**
Related to #357 
Related to #263 

This adds detailed documentation of the proposed new API (with minor changes from the previous version). I've added a couple references for those interested in the design of REST-ish APIs.

Also adds image of new data-model, though this may still change slightly in the near future.

### CODE REVIEW

File: docs/current_api.md [hunk 1]
```diff
@@ -60,7 +57,12 @@ This design is based on moving towards an API that is more RESTful, and attempti
 ### Error Response
 All non-successful responses should return appropriate status codes in the range `4xx` or `5xx`, with `"result": null`, and a descriptive message in `"error": { "message": "..." }`. If the error is due to validation, the `"error"` object should have a key of `"validation"` which is an array of strings describing the validation errors.
 
-### Routes
+### Accept and Content-Type Headers
+Clients should set the `Accept` header in order to indicate what content-type(s) they expect as a response. If no `Accept` header is set, the API will default to a `application/json` content-type. The API will set the `Content-Type` header accurately to reflect the response format. The two response formats that should be supported are:
+* JSON
+* XLS (via XML)
```
Comment (astroash) [inline docs/current_api.md]: Can this be changed to CSV

Comment (RogeredBacon) [reply docs/current_api.md]: Does this relate to the spreadsheets or http requests?

**Comment by eliascodes** [comment]
@astroash 

> Also, are these all the api doc changes you imaged? When you described it I assumed it was going to be bigger.

I suspect you missed the new file I added because the diff was suppressed. Check again.

File: docs/new_api.md [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+For example, in `GET /api/visitors`, the response body is:
+```
+{ result: [{ id, name, gender, email, ... }] }
+```
+We should specify a filter on `gender` and `email` in the following way:
+```
+GET /visitors?gender=foo&email=bar
+```
+
+In addition, the following query parameters should be standard:
```
Comment (astroash) [inline docs/new_api.md]: When you stay they're standard does this mean that every get request should have all of these params?

File: docs/new_api.md [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+* `fields`: Comma separated list of fields in the response body to return. For example, `GET /visitors?fields=name,gender` returns only the `name` and `gender` field in the response body, instead of the entire object.
+* `sort`: Single field by which to sort results
+* `order`: Order by which to sort field specified by `sort`. One of `asc` or `desc`
+* `since`: ISO86001 date string (with timezone?) specifying the **earliest** time from which results should be returned. This is not always meaningful for every endpoint, and is therefore ignored in those cases.
+* `until`: ISO86001 date string (with timezone?) specifying the **latest** time from which results should be returned. This is not always meaningful for every endpoint, and is therefore ignored in those cases.
```
Comment (astroash) [inline docs/new_api.md]: ISO 8601 includes a timestamp with a  'z' offset of UCT.

File: docs/new_api.md [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+}
+```
+
+## Error Response
+All non-successful responses should return appropriate status codes in the range `4xx` or `5xx`, with `"result": null`, and a descriptive message in `"error": { "message": "..." }`. If the error is due to validation, the `"error"` object should have a key of `"validation"` which is an array of strings describing the validation errors.
+
+## Accept and Content-Type Headers
+Clients should set the `Accept` header in order to indicate what content-type(s) they expect as a response. If no `Accept` header is set, the API will default to a `application/json` content-type. The API will set the `Content-Type` header accurately to reflect the response format. The two response formats that should be supported are:
+* JSON
+* XLS (via XML)
```
Comment (astroash) [inline docs/new_api.md]: Change to CSV

File: docs/new_api.md [hunk 5]
```diff
@@ -0,0 +1,646 @@
...
+
+* `NULL`: Un-authenticated user
+* `CB`: Community business user
+* `CB_ADMIN`: Community business user with elevated admin permissions
+* `TWINE_ADMIN`: Twine platform admin user
+
+The `CB_ADMIN` permission level is a short-lived authorization with broader rights than the `CB` permission level. It is attained via a permission escalation mechanism (currently confirming password).
+
+## Activities
+### `GET    /api/activities/`
```
Comment (astroash) [inline docs/new_api.md]: I'm not clear on this layout. Is this meant to be a request to this endpoint without params by a CB?

File: docs/new_api.md [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+    ...
+  },
+  gender: {
+    <GENDER_NAME>: [{ id: String, activity: String, visit_date: String }],
+    ...
+  },
+}
+```
+
+## Design Resources
```
Comment (astroash) [inline docs/new_api.md]: 😍

File: docs/current_api.md (hunk 1 above)
Comment (astroash) [reply docs/current_api.md]: ☝️

File: docs/new_api.md (hunk 2 above)
Comment (eliascodes) [reply docs/new_api.md]: It means that you should be able to pass these parameters to every endpoint and for them to have the desired effect if the parameter is meaningful for that endpoint.

File: docs/new_api.md (hunk 5 above)
Comment (eliascodes) [reply docs/new_api.md]: It's meant to be a request to that endpoint by a client that has a token corresponding to a `CB` role.

File: docs/new_api.md (hunk 2 above)
Comment (astroash) [reply docs/new_api.md]: Surely it's not necessary to write functionality for this is on all `GET` requests. For example on the feedback request will never need to be sorted as there is no tabulated data.

Comment (eliascodes) [reply docs/new_api.md]: We write a function to do this once, and can use it everywhere.

File: docs/new_api.md (hunk 3 above)
Comment (astroash) [reply docs/new_api.md]: ie I would remove "(with timezone?)"

**Pull request merged**